### PR TITLE
Default nameservers to an empty array

### PR DESCRIPTION
**Pull request by jhoblitt**

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Manage an ISC DHCP server
 class dhcp (
   Array[String] $dnsdomain = $dhcp::params::dnsdomain,
-  Array[String] $nameservers = ['8.8.8.8', '8.8.4.4'],
+  Optional[Array[String]] $nameservers = ['8.8.8.8', '8.8.4.4'],
```
Comment (ekohl) [inline manifests/init.pp:4]: IMHO we should default to an empty array instead and skip the option when it's empty. You can't go back to undef if there's a default.

Comment (jhoblitt) [reply manifests/init.pp:4]: Indeed. I'm more than happy to remove the default. To be pedantic, do you want to allow `undef` or restrict it to `[]`.

Comment (ekohl) [reply manifests/init.pp:4]: IMHO `Array[String]` or even `Array[Stdlib::Host]`. For this module an empty array is equivalent to undef and and code is easier to deal with when you can assume an array.

Comment (jhoblitt) [reply manifests/init.pp]: That works for me.  I have revised the PR but there is some awkwardness in that if either `dnsupdatekey` or `ddns_updates` is set, `dhcpd.conf.erb` expects that `dnsupdateserver_real` will have a real value.

File: spec/acceptance/nameservers_spec.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  service_name = case fact('osfamily')
+                 when 'Debian'
+                   'isc-dhcp-server'
+                 else
+                   'dhcpd'
+                 end
+
+  let(:pp) do
+    <<-EOS
+    $interface = $facts['networking']['interfaces'][#{interface}]
```
Comment (ekohl) [inline spec/acceptance/nameservers_spec.rb:21]:
```suggestion
    $interface = $facts['networking']['interfaces']['#{interface}']
```

**Comment by jhoblitt** [comment]
I don't use ddns, so my opinion may not be relevant, but I'm leaning towards `dnsupdateserver` being mandatory to configure ddns functionality.  A less drastic option would be fail if `$dnsupdateserver_real` is `undef`.

**Comment by ekohl** [comment]
> I don't use ddns, so my opinion may not be relevant, but I'm leaning towards `dnsupdateserver` being mandatory to configure ddns functionality. A less drastic option would be fail if `$dnsupdateserver_real` is `undef`.

I agree.

**Comment by jhoblitt** [comment]
Which should I implement in this PR? fail()?

**Comment by ekohl** [comment]
Let's stay compatible and fail only when `$dnsupdateserver_real` is empty.

**Comment by jhoblitt** [comment]
@ekohl I think this PR should be ready for a re-review.  I have added a `fail()` and rspec-puppet + acceptance tests for the error.

**Comment by jhoblitt** [comment]
@ekohl Would you object to also removing the default value for `dnsdomain`?

File: manifests/init.pp [hunk 3]
```diff
@@ -60,7 +60,12 @@
     $bootp_real = $bootp
   }
 
-  $dnsupdateserver_real = pick($dnsupdateserver, $nameservers[0])
+  $dnsupdateserver_real = pick_default($dnsupdateserver, $nameservers[0])
+  if $ddns_updates or $dnsupdatekey {
+    if $dnsupdateserver_real == '' {
```
Comment (ekohl) [inline manifests/init.pp:65]: I think you may be missing undef here. How about:
```suggestion
    unless $dnsupdateserver_real =~ String[1] {
```

**Comment by jhoblitt** [comment]
@ekohl I have a use case where I don't want to set a domain name or nameservers for clients within a pool and I can't seem to do that if there is a global config.

**Comment by ekohl** [comment]
Not against it, but can't you globally set it to an empty array?

**Comment by jhoblitt** [comment]
Uh, looks like the template should work with an empty domain list. I got hung up on dealing with nameservers first.

**Comment by jhoblitt** [comment]
@ekohl Is anything else need to get this PR merged?

**Comment by ekohl** [comment]
Thanks!

**Pull request merged**